src/Algebra/Graph/AdjacencyMap/Algorithm.hs

```diff
@@ -321,20 +326,100 @@ isAcyclic = isRight . topSort
 -- 'isAcyclic' x     == (scc x == 'gmap' NonEmpty.'NonEmpty.vertex' x)
 -- @
 scc :: Ord a => AdjacencyMap a -> AdjacencyMap (NonEmpty.AdjacencyMap a)
-scc m = gmap (component Map.!) $ removeSelfLoops $ gmap (leader Map.!) m
+scc g = condense g $ execState (gabowSCC g) initialState where
+  initialState = C 0 0 [] [] Map.empty Map.empty IntMap.empty [] []
+
+data StateSCC a
+  = C { current       :: !Int
```

Comment:
Please correct me if I'm wrong, but I think GHC doesn't automatically unpack strict fields, which means this (and the next) `Int` will still be represented by a pointer that is guaranteed to point to an evaluated integer.

If my understanding is correct, I think one could use `{-# UNPACK #-}` here to ask GHC to unpack the contents of the field `current` into the constructor `C` itself, removing one level of indirection.

https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#unpack-pragma

Perhaps, this doesn't matter in this case? Then it's fine to leave the code as is but I'm curious to know if this is the case.

Reply:
Ah yes, it probably doesn't matter much (it's a very small part of the work), but I figured it couldn't hurt to add them.